libpod/service.go

```diff
@@ -59,52 +60,84 @@ func (c *Container) IsService() bool {
 	return c.config.IsService
 }
 
+// serviceContainerReport bundles information when checking whether a service
+// container can be stopped.
+type serviceContainerReport struct {
+	// Indicates whether the service container can be stopped or not.
+	canBeStopped bool
+	// Number of all known containers below the service container.
+	numContainers int
+	// Number of containers below the service containers that exited
+	// non-zero.
+	failedContainers int
+}
+
 // canStopServiceContainerLocked returns true if all pods of the service are stopped.
 // Note that the method acquires the container lock.
-func (c *Container) canStopServiceContainerLocked() (bool, error) {
+func (c *Container) canStopServiceContainerLocked() (*serviceContainerReport, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	if err := c.syncContainer(); err != nil {
-		return false, err
+		return nil, err
 	}
 
 	if !c.IsService() {
-		return false, fmt.Errorf("internal error: checking service: container %s is not a service container", c.ID())
+		return nil, fmt.Errorf("internal error: checking service: container %s is not a service container", c.ID())
 	}
 
 	return c.canStopServiceContainer()
 }
 
 // canStopServiceContainer returns true if all pods of the service are stopped.
 // Note that the method expects the container to be locked.
-func (c *Container) canStopServiceContainer() (bool, error) {
+func (c *Container) canStopServiceContainer() (*serviceContainerReport, error) {
+	report := serviceContainerReport{canBeStopped: true}
 	for _, id := range c.state.Service.Pods {
 		pod, err := c.runtime.LookupPod(id)
 		if err != nil {
 			if errors.Is(err, define.ErrNoSuchPod) {
 				continue
 			}
-			return false, err
+			return nil, err
 		}
 
 		status, err := pod.GetPodStatus()
 		if err != nil {
 			if errors.Is(err, define.ErrNoSuchPod) {
 				continue
 			}
-			return false, err
+			return nil, err
 		}
 
-		// We can only stop the service if all pods are done.
+	loop:
 		switch status {
 		case define.PodStateStopped, define.PodStateExited, define.PodStateErrored:
+			podCtrs, err := c.runtime.state.PodContainers(pod)
+			if err != nil {
+				return nil, err
+			}
+			for _, pc := range podCtrs {
+				exitCode, err := c.runtime.state.GetContainerExitCode(pc.ID())
+				if err != nil {
+					return nil, err
+				}
+				if exitCode != 0 {
+					report.failedContainers++
+				}
+				if !pc.IsInfra() {
```

Comment:
Shouldn't this condition wrap also getting `exitCode` and  counting `failedContainers`?